core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -658,7 +658,7 @@ boolean adjustSplitBatchSize(List<SplitAssignment> splitAssignments, long reques
                 splitBatchSize.set(newSplitBatchSize);
             }
             // abandon current request and reschedule update if size of request body exceeds requestSizeLimit and splitBatchSize is updated
-            if (newSplitBatchSize < currentSplitBatchSize) {
+            if (newSplitBatchSize < currentSplitBatchSize && numSplits > newSplitBatchSize) {
```

Comment:
is this not to reschedule when `numSplits <= guaranteedSplitsPerRequest` already?

Reply:
Yep, or any other situation where the new batch size would still allow sending the number of splits that we have for the current request.

Reply:
I think that would be a bug(other than `numSplits <= guaranteedSplitsPerRequest`). Because if `requestSize > maxRequestSizeInBytes` and `newSplitBatchSize < currentSplitBatchSize`, it should be `numSplits <= guaranteedSplitsPerRequest || numSplits > newSplitBatchSize`. Otherwise, that would mean that the split batch size decreasing logic is not working as expected.

Reply:
I believe you're correct, but it seems more clear to make the check explicit and also handle `numSplits < guaranteedSplitsPerRequest` no? I'm happy to just drop this commit since the first one is already merged with the bugfix too.

Reply:
I think this is useful. My thought is that it would be better to have something like `checkState(numSplits > newSplitBatchSize || numSplits <= guaranteedSplitsPerRequest, "numSplits is not adjusted");` inside `if (newSplitBatchSize < currentSplitBatchSize) { ... }`.

Reply:
Ok, updated with that additional check.

Reply:
Actually, we were wrong about that `numSplits` assumption: the new batch size can be smaller than the old batch size even though `numSplits <= newSplitBatchSize`. I've pushed a new revision which updates the check to only reschedule when `numSplits > newSplitBatchSize && requestSize > maxRequestSizeInBytes` since that's logically the intention here.

Reply:
> the new batch size can be smaller than the old batch size even though numSplits <= newSplitBatchSize.
I see that this can happen when `requestSize < expectedSize && currentSplitBatchSize < maxUnacknowledgedSplits` and `currentSplitBatchSize > newSplitBatchSize`. Good catch.